File: src/types/tmux.ts
/**
 * tmuxセッション
 *
 * SSH経由で取得したtmuxセッション情報を表す。
 */
export interface TmuxSession {
  /** セッション名 (unique per server) */
  name: string;
  /** 作成日時 (Unix timestamp ms) */
  created: number;
  /** 他クライアントがアタッチ中か */
  attached: boolean;
  /** ウィンドウ数 */
  windowCount: number;
  /** 所属ウィンドウ (lazy load) */
  windows: TmuxWindow[];
}

/**
 * tmuxウィンドウ
 */
export interface TmuxWindow {
  /** ウィンドウインデックス (0-based) */
  index: number;
  /** ウィンドウ名 */
  name: string;
  /** アクティブウィンドウか */
  active: boolean;
  /** ペイン数 */
  paneCount: number;
  /** 所属ペイン (lazy load) */
  panes: TmuxPane[];
}

/**
 * tmuxペイン
 */
export interface TmuxPane {
  /** ペインインデックス (0-based) */
  index: number;
  /** ペインID (%0, %1, etc.) */
  id: string;
  /** アクティブペインか */
  active: boolean;
  /** 現在実行中のコマンド */
  currentCommand: string;
  /** ペインタイトル */
  title: string;
  /** 幅（カラム数） */
  width: number;
  /** 高さ（行数） */
  height: number;
  /** カーソルX位置 */
  cursorX: number;
  /** カーソルY位置 */
  cursorY: number;
}

/**
 * ペインターゲット指定子
 */
export interface PaneTarget {
  session: string;
  windowIndex: number;
  paneIndex: number;
}

/**
 * ペインターゲットを文字列に変換する
 * @example formatPaneTarget({ session: "main", windowIndex: 0, paneIndex: 1 }) // "main:0.1"
 */
export function formatPaneTarget(target: PaneTarget): string {
  return `${target.session}:${target.windowIndex}.${target.paneIndex}`;
}

File: flutter/lib/models/tmux.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'tmux.freezed.dart';
part 'tmux.g.dart';

/// tmuxセッション
@freezed
class TmuxSession with _$TmuxSession {
  const factory TmuxSession({
    required String name,
    required DateTime created,
    required bool attached,
    required int windowCount,
    @Default([]) List<TmuxWindow> windows,
  }) = _TmuxSession;

  factory TmuxSession.fromJson(Map<String, dynamic> json) =>
      _$TmuxSessionFromJson(json);
}

/// tmuxウィンドウ
@freezed
class TmuxWindow with _$TmuxWindow {
  const factory TmuxWindow({
    required int index,
    required String name,
    required bool active,
    required int paneCount,
    @Default([]) List<TmuxPane> panes,
  }) = _TmuxWindow;

  factory TmuxWindow.fromJson(Map<String, dynamic> json) =>
      _$TmuxWindowFromJson(json);
}

/// tmuxペイン
@freezed
class TmuxPane with _$TmuxPane {
  const factory TmuxPane({
    required int index,
    required String id,
    required bool active,
    required String currentCommand,
    required String title,
    required int width,
    required int height,
    required int cursorX,
    required int cursorY,
  }) = _TmuxPane;

  factory TmuxPane.fromJson(Map<String, dynamic> json) =>
      _$TmuxPaneFromJson(json);
}

/// tmuxセッション拡張
extension TmuxSessionExtension on TmuxSession {
  /// アクティブなウィンドウを取得
  TmuxWindow? get activeWindow {
    try {
      return windows.firstWhere((w) => w.active);
    } catch (_) {
      return windows.isNotEmpty ? windows.first : null;
    }
  }
}

/// tmuxウィンドウ拡張
extension TmuxWindowExtension on TmuxWindow {
  /// アクティブなペインを取得
  TmuxPane? get activePane {
    try {
      return panes.firstWhere((p) => p.active);
    } catch (_) {
      return panes.isNotEmpty ? panes.first : null;
    }
  }
}
